src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs

```diff
@@ -329,83 +329,25 @@ internal static void SaveStreamToFile(Stream stream, string filePath, PSCmdlet c
-            }
-
-            char[] chars = new char[useBufferSize];
-            byte[] bytes = new byte[useBufferSize * 4];
-            int bytesRead = 0;
-            do
-            {
-                // Read at most the number of bytes that will fit in the input buffer. The
-                // return value is the actual number of bytes read, or zero if no bytes remain.
-                bytesRead = stream.Read(bytes, 0, useBufferSize * 4);
-
-                bool completed = false;
-                int byteIndex = 0;
-
-                while (!completed)
-                {
-                    // If this is the last input data, flush the decoder's internal buffer and state.
-                    bool flush = (bytesRead == 0);
-                    decoder.Convert(bytes, byteIndex, bytesRead - byteIndex,
-                                    chars, 0, useBufferSize, flush,
-                                    out int bytesUsed, out int charsUsed, out completed);
-
-                    // The conversion produced the number of characters indicated by charsUsed. Write that number
-                    // of characters to our result buffer
-                    result.Append(chars, 0, charsUsed);
-
-                    // Increment byteIndex to the next block of bytes in the input buffer, if any, to convert.
-                    byteIndex += bytesUsed;
-
-                    // The behavior of decoder.Convert changed start .NET 3.1-preview2.
-                    // The change was made in https://github.com/dotnet/coreclr/pull/27229
-                    // The recommendation from .NET team is to not check for 'completed' if 'flush' is false.
-                    // Break out of the loop if all bytes have been read.
-                    if (!flush && bytesRead == byteIndex)
-                    {
-                        break;
-                    }
-                }
-            } while (bytesRead != 0);
-
-            return result.ToString();
-        }
-
         internal static string DecodeStream(Stream stream, string characterSet, out Encoding encoding)
         {
-            bool isDefaultEncoding = false;
-            byte[] buffer = new byte[4];
+            bool isDefaultEncoding = !TryGetEncodingFromCharset(characterSet, out encoding);
 
-            if (stream.Length >= 4)
-            {
-                stream.ReadExactly(buffer, 0, 4);
-            }
+            using StreamReader reader = new(stream, encoding, detectEncodingFromByteOrderMarks: true);
 
-            EncodingHelper.TryDetectEncodingFromBom(buffer, out encoding, out int preambleLength);
-
-            if (encoding is null)
-            {
-                isDefaultEncoding = !TryGetEncodingFromCharset(characterSet, out encoding);
-            }
-
-            stream.Seek(preambleLength, SeekOrigin.Begin);
-            string content = StreamToString(stream, encoding);
+            encoding = reader.CurrentEncoding;
```

Comment:
@iSazonov, I'm afraid I don't understand the question (I'm not familiar with the code).

Reply:
@mklement0 I ask about p.3 from https://github.com/PowerShell/PowerShell/issues/11547#issuecomment-636522169
> otherwise, for XML and HTML, respect the encoding specified in the XML declaration

Reply:
@mklement0 Friendly ping.

Reply:
@iSazonov, sorry, I forgot to respond: I'm still unclear on the details, but what I suggested in https://github.com/PowerShell/PowerShell/issues/11547#issuecomment-636522169 meant looking no further once a BOM is found. Only if there's none, look at `charset`. Only if there's none, look at the XML declaration / HTML `<meta>` element (the latter seems to be absent from the snippet above - is HTML handled elsewhere?). 
This hinges on explicitly knowing if a BOM is present.

Reply:
@mklement0 Thanks! What I don't understand is how it should work if the encoding is specified inside the XML/HTML. Now we always perform stream conversion to UTF16 string first, which means subsequent conversion based on attribute from XML will not be correct, right? That means we either have to ignore that XML attribute or change our decoding completely and do it on the original byte stream.

Reply:
(As for also looking for a `<meta>` element: I didn't read carefully the first time, and I now see that it is also being detected).

Yes, there is a problem with the detection code: it only works properly if the actual encoding is a _single-byte_ encoding that is a _superset_ of ASCII, so it only works with the various OEM and ANSI single-byte Windows encodings and UTF-8, for instance, but _not_ with other Unicode encodings (UTF-16LE, UTF-16BE, ...)

While limiting the detection to ASCII supersets strikes me as sufficient in practice (I don't know what the RFCs say),
additional work is required to detect the other Unicode encodings, which, however would be as simple as doing `substring.Replace("\0", "")` before calling `.Match(substring)` on the regexes, based on how those multi-byte Unicode encodings encode ASCII-range characters.

Aside from that, as noted, the current code also doesn't honor the proposed precedence rules.  (I am aware that its fundamentals predate your modifications, @CarloToso.) 

If we had explicit BOM detection, we could then simply skip the regex matching if a BOM is present.

Reply:
@mklement0 Thanks! So simplest fix is TryConvertToXml should use raw stream instead of string so that we delegate decoding to XmlReader, right?

As for precedence rules, if we agree to decode XML from raw stream this will resolve this too.

Reply:
Good point; delegating to `System.Xml.XmlDocument.Load(System.IO.Stream inStream)` with a raw byte stream is an option, but the above is still necessary for an HTML `<meta>` element.

However, this would result in _difference_ precedence rules, because the file/raw-bytes `.Load()` overload actually gives precedence to the `encoding` attribute in an XML declaration over a BOM - even though that can _break_ the parsing (e.g., if a single-byte encoding specified via `charset` conflicts with a multi-byte BOM).

In light of all this, perhaps it's better to change the precedence rules to prioritize explicit _out-of-band_ information, i.e. a `charset` spec. in the `Content-Type` header field, as follows, across all mediate types:

* Look for a (valid) `charset` attribute in the `Content-Type` header field, and, if present, honor it - no further investigation needed.

* Examine the first four raw bytes to see if they constitute a BOM - see https://en.wikipedia.org/wiki/Byte_order_mark#Byte_order_marks_by_encoding; restricting this to UTF-8 and the UTF-16 and UTF-32 variants is probably sufficient.

  * If so, always honor the specified encoding - and, in the case of XML, use _string_-based XML parsing (in which case the XML declaration's `encoding` attribute does NOT matter) - no further investigation needed.

* If not, look for an encoding specified via `<meta>` elements or XML-declarations and, if found, honor it (in the XML case, the raw-byte `.Load()` overload can then be used) - no further investigation needed.

* If there is none, default to UTF-8

Reply:
Unfortunately if we change the precedence rules we fail to fix https://github.com/PowerShell/PowerShell/issues/11547, charset=utf8, actual encoding=utf8withBom. I think we should check for the BOM and then for the charset

Reply:
@CarloToso, I don't think we need to handle this case, as it amounts to an _over-specification_ that we shouldn't be expect to compensate for:

If `charset=utf8` is present in the `Content-Type` header field _and_ there is (inappropriately) _also_ a UTF-8 BOM, I think it's fine (and appropriate) to parse this BOM as the actual character it represents, i.e. as [`U+FEFF`](http://www.fileformat.info/info/unicode/char/feff)) (ZERO WIDTH NO-BREAK SPACE)

Reply:
Our code resembles scraps. It should be made cleaner by avoiding breaking changes.
So first we call TryProcessFeedStream that is XML parsing of RSS/Feed based on raw stream. Then on fail we try the same thing again but trying to guess the response type and the encoding. It is not clear why we ignore the response type when we try to parse RSS/Feed. I believe we could remove one extra step.

Reply:
@iSazonov, good point about breaking changes.
Note that I am not deeply immersed in this, so do tell me if I'm getting something wrong:

Looking at the old code (https://github.com/PowerShell/PowerShell/blob/420c29d12c65e49a8d76909162286c24eee11d17/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs#L389), I see that the _out-of-band_ information - i.e. a `charset` attribute - currently already takes precedence, short-circuiting further investigation.

So we need to stick with that to avoid a breaking change, but do note that it contradicts the relevant RFC, as noted in https://github.com/PowerShell/PowerShell/issues/11547#issuecomment-637034379
Rethinking my over-specification statement: perhaps the right thing to do, if _both_ out-of-band encoding information and a BOM are present, is to remove the BOM _if the indicated / implied encoding is the same_.

In the absence of out-of-band information, the question is then how the handle the _in-band_ information - BOM and/or `encoding` / `charset` attribute in `<?xml>` declaration / `<meta>` HTML element.

The current code searches _only_ for the latter, and does so _incorrectly_, as discussed (no support for 2+-byte Unicode encodings). 

To fix this:

* For _XML_-based in-band encoding information, we can delegate to the .NET APIs based on raw byte streams.
* For the HTML `<meta>` case, we still need to dour own text-based matching, but it requires something like the `.Replace("\0")` approach mentioned above.

This too avoids a breaking change, though (like now) it requires the relatively expensive parsing of the first 1K / letting a .NET method call potentially fail (for potential XML content).

Giving precedence to manual BOM detection would perform better at least with non-XML content, but the only way  this would _not_ amount to a breaking change is _if the presence of a BOM currently breaks things_, which is the case, right?

Manual BOM detection is still necessary, at least in the absence of out-of-band encoding information and possibly also XML-declaration / HTML `<meta>` in-band information.

Reply:
.Net does encoding detection for XML https://source.dot.net/#System.Private.Xml/System/Xml/Core/XmlTextReaderImpl.cs,2895
So we have no need to do the same but worse.
Since first we try Feeds/Atom as XML we can simplify the code and first try XML reading then fallback to Json.

Reply:
@iSazonov we only try Feeds/Atom as XML in `Invoke-RestMethod` not in `Invoke-WebRequest`

Reply:
This does not prevent my suggestion.